### PR TITLE
PHP 8 Support

### PR DESCRIPTION
**Pull request by driesvints**
| Q                 | A
| ----------------- | ----------
| Bug fix?          | no
| New feature?      | yes    
| BC breaks?        | no     
| Related Issue     | Fix #...  
| Need Doc update   | no

## Describe your change

This PR will provide PHP 8 Support for the library. Atm this is blocking https://github.com/laravel/scout/pull/425

I'll keep the PR in draft until I've fixed the build.

### CODE REVIEW

**Comment by julienbourdeau** [comment]
Thanks Dries!
Was it enough to run it on PHP 8 locally?

**Comment by driesvints** [comment]
@julienbourdeau I'll need to verify that on thursday

**Review by devdreamsolution — DISMISSED** [review]

**Comment by driesvints** [comment]
@julienbourdeau I wasn't able to get things running on PHP 8 I'm afraid. I think it's best that we get rid of the PHP CS Fixer dependency as soon as you can. Or at least find a way how we can add `--ignore-platform-req=php` to the nightly build.

**Comment by julienbourdeau** [comment]
I think removing php-cs-fixer in the deps should be fine. It's only needed in the CI anyway

**Comment by driesvints** [comment]
I'm gonna wait a bit with proceeding here until the CircleCI PR has been merged so I can rebase on it.

**Comment by chloelbn** [comment]
@driesvints circleCI has been merged on master! you can keep going. Thanks for your work!

**Comment by driesvints** [comment]
Seems like builds for forked PRs aren't enabled yet: https://circleci.com/docs/2.0/oss/#build-pull-requests-from-forked-repositories

**Comment by chloelbn** [comment]
@driesvints Nice catch thanks! it should be good now

File: CHANGELOG.md [hunk 1]
```diff
@@ -8,6 +8,9 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
 
 ## [Unreleased](https://github.com/algolia/algoliasearch-client-php/compare/2.7.2...master)
 
+### Added
+- PHP 8 Support
+
```
Comment (julienbourdeau) [inline CHANGELOG.md:13]:
```suggestion


```

File: composer.json [hunk 2]
```diff
@@ -21,8 +21,8 @@
     },
     "require-dev": {
         "friendsofphp/php-cs-fixer": "^2.0",
-        "fzaninotto/faker": "^1.8",
-        "phpunit/phpunit": "^4.8",
+        "fakerphp/faker": "^1.9.1",
+        "phpunit/phpunit": "^4.8 || ^9.3",
```
Comment (julienbourdeau) [inline composer.json:25]:
```suggestion
        "phpunit/phpunit": "^4.8",
```

**Comment by julienbourdeau** [comment]
PHP 8.0 doesn't seem available on circle-ci yet: `Error response from daemon: manifest for circleci/php:8.0 not found`.

Our tests were written for PHPUnt 4 and won't work with PHPUnit 9 unfortunately.

**Comment by chloelbn** [comment]
@julienbourdeau luckily for us, as implementing the CTS is in our pipeline we'll be able to upgrade PHPUnit as well too, right?

**Comment by julienbourdeau** [comment]
@chloelbn Unfortunately, PHPUnit 9 is only compatible with recent php version. And since there are breaking change, you can't have one test suite that works in old PHPUnit and newer PHPUnit 😬 
Not sure how we're going to deal with it yet 😩

**Comment by driesvints** [comment]
@julienbourdeau pushed a change to name the builds. They now appear much more readable here in the GitHub checks overview.

We'll need to provide a matrix for the PHPUnit versions.

File: tests/tests-no-composer.php [hunk 3]
```diff
@@ -14,9 +14,10 @@
 
 function safeName($name)
 {
-    if (getenv('CI_BUILD_NUM')) {
-        return sprintf('php_%s_%s', $name, getenv('CI_BUILD_NUM'));
-    }
-
-    return $name;
+    return sprintf(
```
Comment (chloelbn) [inline tests/tests-no-composer.php:17]: this fix seems to make the CI fail

**Review by chloelbn — CHANGES_REQUESTED** [review]

File: tests/tests-no-composer.php [hunk 4]
```diff
@@ -14,9 +14,10 @@
...
+        getenv('CI_BUILD_NUM') ? 'TRAVIS_php_%s_%s_%s' : 'php_%s_%s_%s',
```
Comment (chloelbn) [inline tests/tests-no-composer.php:18]:
```suggestion
        getenv('CI_BUILD_NUM') ? 'CIRCLE_php_%s_%s_%s' : 'php_%s_%s_%s',
```

**Review by chloelbn — CHANGES_REQUESTED** [review]

**Comment by chloelbn** [comment]
@driesvints @julienbourdeau I'm not sure we want to provide a matrix for the PHPUnit version. I fear it will crowd up the queue in the CI. Defining it in `composer.json` is enough IMO. WDYT?

**Comment by driesvints** [comment]
Sure, as long as PHPUnit 4 tests pass on PHP 8. We just need to figure out on how to get a working build for that now. Probably needs a different Docker image.

**Comment by driesvints** [comment]
Converted back to draft until we can test on PHP 8.

CircleCI is providing support for this soon it seems: https://github.com/CircleCI-Public/cimg-php/pull/96

**Comment by felicianotech** [comment]
The CircleCI PHP 8 image (`cimg/php:8.0.0`) is now ready.

**Comment by driesvints** [comment]
@felicianotech thanks. Isn't there a "latest" tag like on circleci/php?

**Comment by driesvints** [comment]
Okay I've managed to get the builds working properly now. I've migrated everything to cimg/php except for the 7.0 build for which there isn't a tag on cimg/php. 

I'm currently stuck at this error:

```
PHP Fatal error:  Uncaught Algolia\AlgoliaSearch\Exceptions\BadRequestException: Index not allowed with this API key in /home/circleci/project/src/RetryStrategy/ApiWrapper.php:219
Stack trace:
#0 /home/circleci/project/src/RetryStrategy/ApiWrapper.php(167): Algolia\AlgoliaSearch\RetryStrategy\ApiWrapper->handleResponse()
#1 /home/circleci/project/src/RetryStrategy/ApiWrapper.php(101): Algolia\AlgoliaSearch\RetryStrategy\ApiWrapper->request()
#2 /home/circleci/project/src/SearchIndex.php(111): Algolia\AlgoliaSearch\RetryStrategy\ApiWrapper->write()
#3 /home/circleci/project/tests/tests-no-composer.php(12): Algolia\AlgoliaSearch\SearchIndex->setSettings()
#4 {main}
  thrown in /home/circleci/project/src/RetryStrategy/ApiWrapper.php on line 219
```

It seems there's something wrong with the Algolia keys but I'm not sure what. Can any of you figure that one out?

In the meantime, the PHP 8 build is running well but it's failing because of an old PHPUnit version: 

```
PHP Fatal error:  Uncaught Error: Call to undefined function each() in /home/circleci/project/vendor/phpunit/phpunit/src/Util/Getopt.php:38
Stack trace:
#0 /home/circleci/project/vendor/phpunit/phpunit/src/TextUI/Command.php(240): PHPUnit_Util_Getopt::getopt()
#1 /home/circleci/project/vendor/phpunit/phpunit/src/TextUI/Command.php(111): PHPUnit_TextUI_Command->handleArguments()
#2 /home/circleci/project/vendor/phpunit/phpunit/src/TextUI/Command.php(100): PHPUnit_TextUI_Command->run()
#3 /home/circleci/project/vendor/phpunit/phpunit/phpunit(52): PHPUnit_TextUI_Command::main()
#4 {main}
  thrown in /home/circleci/project/vendor/phpunit/phpunit/src/Util/Getopt.php on line 38
```

**Comment by driesvints** [comment]
PHPUnit 8.5/9.4 is failing on the void return type as I expected. This leaves us with two options. Either drop everything below PHP 7 or duplicate the test suite for older PHP versions. A third option is to not build on PHP 5.4 anymore.

**Comment by driesvints** [comment]
There's a fourth option we can take maybe. PHP Webdriver patches their tests for newer PHPUnit versions with this script: https://github.com/php-webdriver/php-webdriver/pull/828/files#diff-bde422e37af67712869f8827f3dbc90538ab9fe13556cf32bd6d3997ba05ca4aR1-R20

We could do the same and then we'd at least be able to see what fails on PHP 8.

**Comment by driesvints** [comment]
A fifth option which is maybe a very reasonable one is to create a new package for this client where you only support modern PHP versions. That's a very reasonable choice given that we're currently only blocked by the test suite. The new package could support PHP 7.2 and up?

**Comment by driesvints** [comment]
I found even a sixth option by advice from @crynobone. We could look into adoption https://github.com/Yoast/PHPUnit-Polyfills maybe instead of PHPUnit.

**Comment by chloelbn** [comment]
@driesvints about the index name error, you have to revert the renaming of the `safeName()` method like so
```
function safeName($name)
{
    if (getenv('CI_BUILD_NUM')) {
        return sprintf('php_%s_%s', $name, getenv('CI_BUILD_NUM'));
    }

    return $name;
}
```

We're considering indeed releasing a new major that could drop the older PHP versions. In the meantime however, could you try out option #4 (I really thank you for all the research you're going on this matter!) and see how it works out? it would seem the cleaner short-term solution

**Comment by driesvints** [comment]
Hey @chloelbn. I've reverted the changes to the `safeName` method but the builds still seems to fail: https://app.circleci.com/pipelines/github/algolia/algoliasearch-client-php/97/workflows/1e1a9d2e-c7b8-40e3-85bd-a50edc66b250/jobs/460

**Comment by thadbryson** [comment]
PHP-CS-Fixer added PHP 8 support in version 2.17.0 so that shouldn't be a blocker anymore. 

https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/2.17/composer.json

**Comment by thadbryson** [comment]
I added my own pull request to get unit tests to work. I bumped PHPUnit to ^9.3.

**Comment by driesvints** [comment]
@chloelbn I've tried implementing the script but now the tests are failing on PHP because the old state of the tests don't work on newer PHPUnit versions anymore: https://app.circleci.com/pipelines/github/algolia/algoliasearch-client-php/104/workflows/2f529bfb-a235-402b-afb3-59454828a064/jobs/531

https://github.com/Yoast/PHPUnit-Polyfills seems like an alternative but requires the entire test suite to be rewritten as PHPUnit 9 tests.

At this point I'd suggest to create a separate 2.x branch from master which contains the current v2 releases and do a new v3 major version on master (or separate 3.x branch) which drops everything below PHP 7.2. It's the most viable choice because supporting all these PHP versions in a single branch just isn't feasible or maintainable.

**Comment by thadbryson** [comment]
> @chloelbn I've tried implementing the script but now the tests are failing on PHP because the old state of the tests don't work on newer PHPUnit versions anymore: https://app.circleci.com/pipelines/github/algolia/algoliasearch-client-php/104/workflows/2f529bfb-a235-402b-afb3-59454828a064/jobs/531
> 
> https://github.com/Yoast/PHPUnit-Polyfills seems like an alternative but requires the entire test suite to be rewritten as PHPUnit 9 tests.
> 
> At this point I'd suggest to create a separate 2.x branch from master which contains the current v2 releases and do a new v3 major version on master (or separate 3.x branch) which drops everything below PHP 7.2. It's the most viable choice because supporting all these PHP versions in a single branch just isn't feasible or maintainable.

I agree. They'll have to drop support for PHP 5 and stick with 7.1 and higher. Maintaining both is not feasible.

**Comment by asivaneswaran** [comment]
Just wondering if there was any ETA on this?

This is kinda blocking me to move my project to 8.0.

Thanks

**Comment by lahivee** [comment]
It really isn't a plus for Algolia to fail keeping their repositories up2date even 3 weeks (!) after the latest PHP release.

**Comment by thadbryson** [comment]
I encourage everyone who wants Algolia to bump support to PHP 8.0 to contact them. 

https://www.algolia.com/support/

Currently their documention says they support everything at and above 5.3 which is no longer accurate.

**Comment by chloelbn** [comment]
@thadbryson please do not reach out for support, we have to be mindful of their time while the dedicated API client team is already actively working on this matter. We'll soon be able to merge [this PR](https://github.com/algolia/algoliasearch-client-php/pull/656) from @julienbourdeau.

Indeed a new major is required and we're preparing it as we speak, but to make all the changes required it won't be ready before a few weeks. We'll deliver a quick fix in the meantime through the previous PR so that our client won't be in your way anymore to upgrade to PHP 8.

Thanks for your patience and your understanding 💖

**Comment by chloelbn** [comment]
PHP 8 support has been added to the latest patch version, `2.7.3` 🎉 
I'm closing this PR, thanks a lot @driesvints and everyone else for your work! A new major version is in the pipeline, but for the moment we hope that this new version will help you move forward in your own projects! Please let us know if you encounter any issues.

Cheers!

**Comment by driesvints** [comment]
Thanks @chloelbn & @julienbourdeau 👍

**Comment by asivaneswaran** [comment]
Great news, thanks everyone involved!

@chloelbn @driesvints @julienbourdeau

**Comment by arcticlinux** [comment]
Thanks everyone!

**Pull request closed without merging**